File: Kotlin/ch13/src/main/kotlin/ch13/LCSDP.kt
package ch13

/**
 * LCS 동적 계획법 + 역추적 (Longest Common Subsequence - DP with traceback)
 * Rust/ch13/src/lcsdp.rs → Kotlin 이디엄 포팅
 */
fun lcsdp(x: ByteArray, y: ByteArray, table: Array<IntArray>): Int {
    for (m in 0..x.size) table[m][0] = 0
    for (n in 0..y.size) table[0][n] = 0
    for (m in 1..x.size) {
        for (n in 1..y.size) {
            table[m][n] = when {
                x[m - 1] == y[n - 1]          -> table[m - 1][n - 1] + 1
                table[m][n - 1] >= table[m - 1][n] -> table[m][n - 1]
                else                           -> table[m - 1][n]
            }
        }
    }
    return table[x.size][y.size]
}

fun lcsTraceBack(x: ByteArray, y: ByteArray, m: Int, n: Int, table: Array<IntArray>, result: StringBuilder) {
    if (m == 0 || n == 0) return
    when {
        table[m][n] > table[m][n - 1] && table[m][n] > table[m - 1][n] && table[m][n] > table[m - 1][n - 1] -> {
            lcsTraceBack(x, y, m - 1, n - 1, table, result)
            result.append(x[m - 1].toInt().toChar())
        }
        table[m][n] > table[m - 1][n] && table[m][n] == table[m][n - 1] ->
            lcsTraceBack(x, y, m, n - 1, table, result)
        else ->
            lcsTraceBack(x, y, m - 1, n, table, result)
    }
}

fun main() {
    val x = "GOOD MORNING.".toByteArray()
    val y = "GUTEN MORGEN.".toByteArray()
    val table = Array(x.size + 1) { IntArray(y.size + 1) }
    val len = lcsdp(x, y, table)
    val sb = StringBuilder()
    lcsTraceBack(x, y, x.size, y.size, table, sb)
    println("LCS length: $len, LCS: $sb")   // 7, "G MORN."
}

File: Rust/ch13/src/bin/lcsdp.rs
// Longest Common Subsequence (Dynamic Programming, with traceback) - 로직은 ch13::lcsdp 모듈에 있다.
// 버그 수정 내용은 ch13::lcsdp 모듈과 bug-fix-report.md 참조.
use ch13::lcsdp::{lcs, lcs_print_table, lcs_trace_back};

fn main() {
    let x = b"GOOD MORNING.";
    let y = b"GUTEN MORGEN.";

    let len_x = x.len();
    let len_y = y.len();

    let mut table = vec![vec![0i32; len_y + 1]; len_x + 1];

    let length = lcs(x, y, len_x, len_y, &mut table);

    lcs_print_table(&table, x, y, len_x, len_y);

    let mut result = String::new();
    lcs_trace_back(x, y, len_x, len_y, &table, &mut result);

    println!();
    println!("LCS:\"{}\" (Length:{})", result, length);
}
